### DISCUSSION TITLE
Nullable ValueTuple feedback

### DISCUSSION THREAD

**Discussion by gafter** [dotnet/csharplang#479] [General]
@jcouv commented on [Fri Apr 21 2017](https://github.com/dotnet/roslyn/issues/18885)

@niemyjski commented on [Fri Apr 21 2017](https://github.com/dotnet/corefx/issues/18745)

I get that this is a weird use case, but it's something the compiler allows me to do and I was even prompted by autocomplete to make my ValueTuple (a struct) nullable...

I was playing around with value tuples and in some cases I needed to return a null value (to say I can't handle it via a plugin). It was at this point autocomplete told me to make it nullable and did so for me. Granted this is the exact point in time, I should have created a class for this... But I was like ooh I can still have the pretty syntax and it works... I noticed that this quickly became a nightmare once I needed to handle null. I guess I was expecting destructuring to work and make my values null (if possible, or throw an compiler error if the tuple item type didn't support it)

```cs
        static void Main(string[] args)
        {
            (string subject, object data)? GetData() {
                return null;
            }

            /*
            Error	CS8130	Cannot infer the type of implicitly-typed deconstruction variable 'data'.
            Error	CS1061	'(string subject, object data)?' does not contain a definition for 'Deconstruct' and no extension method 'Deconstruct' accepting a first argument of type '(string subject, object data)?' could be found (are you missing a using directive or an assembly reference?)
            */
            (var subject, var data) = GetData();

            /*
            Error	CS1061	'(string subject, object data)?' does not contain a definition for 'Deconstruct' and no extension method 'Deconstruct' accepting a first argument of type '(string subject, object data)?' could be found (are you missing a using directive or an assembly reference?)
            Error	CS8129	No Deconstruct instance or extension method was found for type '(string subject, object data)?', with 2 out parameters and a void return type.
             */
            (string subject, object data) = GetData();

            // works
            var result = GetData();
            if (result == null || result.Value.subject == null)
                return;

            // Also I get that ValueTuple is a mutable struct.... This should work no?
            //Error CS1612  Cannot modify the return value of '(string subject, object data)?.Value' because it is not a variable
            result.Value.subject = result.Value.subject.ToLower();
        }
```

---

@niemyjski commented on [Fri Apr 21 2017](https://github.com/dotnet/corefx/issues/18745#issuecomment-296268670)

cc @jcouv

---

@jcouv commented on [Fri Apr 21 2017](https://github.com/dotnet/roslyn/issues/18885#issuecomment-296276019)

@VSadov The last error (`CS1612`) surprised me. Does it make any sense to you? If not, I'll investigate.

---

@svick commented on [Fri Apr 21 2017](https://github.com/dotnet/roslyn/issues/18885#issuecomment-296278348)

@jcouv I think it's expected that if a property returns a value type, you can't modify the returned temporary value directly (because any changes would be lost).

---

@svick commented on [Fri Apr 21 2017](https://github.com/dotnet/roslyn/issues/18885#issuecomment-296279842)

> I guess I was expecting destructuring to work and make my values null

I don't think that would be a good idea. If you want to return a tuple that contains `null`s, you can:

```c#
(string subject, object data) GetData() {
    return (null, null);
}
```

But you decided to not do this and instead use nullable type. I think the compiler shouldn't ignore you made that decision.

---

@niemyjski commented on [Fri Apr 21 2017](https://github.com/dotnet/roslyn/issues/18885#issuecomment-296281371)

@svick doh, I didn't even think of that... I just did return null not thinking (and was prompted to make it nullable) and then everything hit the fan lol

---

@dasMulli commented on [Fri Apr 21 2017](https://github.com/dotnet/roslyn/issues/18885#issuecomment-296289180)

Two workarounds:

1. Use `??`: `var (subject, data) = GetData() ?? (null, null);`

2. Create an extension method for `ValueTuple<T1,T1>?`:

```c#
public static void Deconstruct<T1,T2>(this ValueTuple<T1,T2>? tuple, out T1 item1, out T2 item2)
{
    if (tuple.HasValue)
    {
        item1 = tuple.Value.Item1;
        item2 = tuple.Value.Item2;
    }
    else
    {
        item1 = default(T1);
        item2 = default(T2);
     }
}
```

While (2) seems tempting, using `default(T1)` would mean `0` for ints for example so I feel this shouldn't be included in the NuGet. (Or maybe with some generic constraints)

---

@gafter commented on [Fri Apr 21 2017](https://github.com/dotnet/roslyn/issues/18885#issuecomment-296312806)

If I understand correctly, I think this is by design.

**Comment by niemyjski** [comment]
This might be the most moved issue on GitHub 🤣
